File: ios/Share/ShareVC.m
//
// ShareViewController.m
// Share
//
// Created by Marcus Rohrmoser on 18.03.15.
// Copyright (c) 2015-2016 Marcus Rohrmoser http://mro.name/me. All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#import "ShareVC.h"
#import <MobileCoreServices/UTCoreTypes.h>
#import "ShaarliM.h"
#import "ShaarliCmdPost.h"


static inline NSString *stringFromPrivacy(const BOOL priv)
{
    return priv ? NSLocalizedString(@"Private 🔐", @"ShaareVC") : NSLocalizedString(@"Public 🔓", @"ShaareVC");
}


static inline const BOOL privacyFromString(NSString *s)
{
    return ![stringFromPrivacy (NO) isEqualToString:s];
}


@interface ShareVC() <UITextFieldDelegate, UITextViewDelegate, ShaarliCmdPostDelegate> {
    SLComposeSheetConfigurationItem *itemTitle;
    SLComposeSheetConfigurationItem *itemAudience;
}
@property (readonly, strong, nonatomic) ShaarliM *shaarli;
@property (readwrite, strong, nonatomic) ShaarliCmdPost *post;
@property (readwrite, assign, nonatomic) BOOL wasTouched;

@property (readwrite, strong, nonatomic) NSMutableDictionary *postForm; // maybe move to ShaarliCmdPost ?
@property (readwrite, strong, nonatomic) NSURL *postURL; // maybe move to ShaarliCmdPost ?
@end


@implementation ShareVC

-(void)viewDidLoad
{
    MRLogD(@"-", nil);
    [super viewDidLoad];
    if( !self.shaarli ) {
        _shaarli = [[ShaarliM alloc] init];
        [self.shaarli load];
    }
}


-(NSArray *)configurationItems
{
    MRLogD(@"-", nil);
    itemTitle = [[SLComposeSheetConfigurationItem alloc] init];
    itemTitle.title = NSLocalizedString(@"Title", @"ShaareVC");
    itemTitle.value = self.contentText;

    itemAudience = [[SLComposeSheetConfigurationItem alloc] init];
    itemAudience.title = NSLocalizedString(@"Audience", @"ShaareVC");
    itemAudience.value = stringFromPrivacy(NO);
    __weak typeof(itemAudience) wr = itemAudience;
    __weak typeof(self) weakSelf = self;

    [itemAudience setTapHandler:^(void) {
         wr.value = stringFromPrivacy ( !privacyFromString (wr.value) );
         weakSelf.wasTouched = YES;
     }
    ];

    return @[itemTitle, itemAudience];
}


-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.view.tintColor = [UIColor colorWithRed:128 / 255.0f green:173 / 255.0f blue:72 / 255.0f alpha:1.0f];
    NSParameterAssert(self.shaarli);
    NSParameterAssert(itemTitle);

    self.title = self.shaarli.title;
    self.textView.keyboardType = UIKeyboardTypeTwitter;

    self.wasTouched = NO;
    itemTitle.value = self.contentText;
    NSString *tagsDefault = self.shaarli.tagsActive ? self.shaarli.tagsDefault : @"";
    {
        NSString *txt = @"🔄";
        if( 0 < tagsDefault.length )
            txt = [[tagsDefault stringByAppendingString:@" "] stringByAppendingString:txt];
        self.textView.text = txt;
    }
    itemAudience.value = stringFromPrivacy(self.shaarli.privateDefault);

    if( !self.shaarli.isSetUp )
        return;
    NSParameterAssert(self.shaarli.isSetUp);
    ShaarliCmdPost *re = [[ShaarliCmdPost alloc] init];
    re.session = self.shaarli.postSession;
    re.endpointURL = self.shaarli.endpointURL;
    re.delegate = self;
    self.post = re;

    for( NSExtensionItem *item in self.extensionContext.inputItems ) {
        for( NSItemProvider *itemProvider in item.attachments ) {
            // see predicate from http://stackoverflow.com/a/27932776
            NSString *t = (NSString *)kUTTypeURL;
            if( [itemProvider hasItemConformingToTypeIdentifier:t] ) {
                [itemProvider loadItemForTypeIdentifier:t options:nil completionHandler:^(NSURL * url, NSError * error) {
                     MRLogD (@"done. title:%@ url:%@ error:%@", itemTitle.value, url, error, nil);
                     if( !error ) {
                         [re startPostForURL:url title:itemTitle.value desc:nil];
                     } else {
                         MRLogW (@"Error: %@", error, nil);
                     }
                 }
                ];
                return;
            }
        }
    }
}


-(void)viewDidAppear:(BOOL)animated
{
    MRLogD(@"-", nil);
    [super viewDidAppear:animated];
    if( !self.shaarli.isSetUp ) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"No Shaarli", @"Share") message:NSLocalizedString(@"There is no Shaarli account configured. You can add one in the Shaarli💫 in-app settings.", @"Share") preferredStyle:UIAlertControllerStyleAlert];
        [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Share") style:UIAlertActionStyleCancel handler:^(UIAlertAction * action) {
                              [self cancel];
                          }
         ]];
        dispatch_async (dispatch_get_main_queue (), ^{
                            [self presentViewController:alert animated:YES completion:nil];
                        }
                        );
        return;
    }
}


-(void)presentationAnimationDidFinish
{
    MRLogD(@"-", nil);
}


-(BOOL)isContentValid
{
    MRLogD(@"-", nil);
    self.wasTouched = YES;
    return YES;
}


-(UIView *)loadPreviewView
{
    return nil;
}


-(void)didSelectPost
{
    MRLogD(@"-", nil);

    NSMutableDictionary *form = self.postForm;
    NSURL *dst = self.postURL;
    NSParameterAssert(form);
    NSParameterAssert(dst);

    [form setValue:itemTitle.value forKey:K_F_LF_TITLE]; // in case of text notes: set it again!
    if( self.shaarli.tagsActive ) {
        NSMutableArray *tags = [NSMutableArray arrayWithCapacity:5];
        [form setValue:[self.contentText stringByStrippingTags:tags] forKey:K_F_LF_DESCRIPTION];
        [form setValue:[tags componentsJoinedByString:@" "] forKey:K_F_LF_TAGS];
    } else
        [form setValue:self.contentText forKey:K_F_LF_DESCRIPTION];
    if( privacyFromString(itemAudience.value) )
        [form setValue:HTML_ON forKey:K_F_LF_PRIVATE];
    else
        [form removeObjectForKey:K_F_LF_PRIVATE];

    [self.post finishPostForm:form toURL:dst];
}


-(void)didSelectCancel
{
    MRLogD(@"-", nil);
    NSError *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSUserCancelledError userInfo:nil];
    [self.extensionContext cancelRequestWithError:error];
    self.post = nil;
}


-(void)cancelWithError:(NSError *)error
{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Shaarlying failed", @"ShareVC") message:[NSString stringWithFormat:NSLocalizedString(@"%@\n\nFailing call was %@", @"ShareVC"), error.localizedDescription, error.userInfo[NSURLErrorKey]] preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"ShareVC") style:UIAlertActionStyleCancel handler:^(UIAlertAction * action) {
                          // http://www.pixeldock.com/blog/ios8-share-extension-completionhandler-for-loaditemfortypeidentifier-is-never-called/
                          // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
                          [super cancel];
                          self.post = nil;
                      }
     ]];
    [self presentViewController:alert animated:YES completion:nil];
}


#pragma mark ShaarliCmdPostDelegate


// Update GUI
-(void)form2Gui:(NSDictionary *)form
{
    itemAudience.value = stringFromPrivacy([HTML_ON isEqualToString:form[K_F_LF_PRIVATE]]);
    NSString *txt = form[K_F_LF_DESCRIPTION];
    if( self.shaarli.tagsActive ) {
        NSString *tags = [form[K_F_LF_TAGS] stringByReplacingOccurrencesOfString:@" " withString:@" #"];
        if( 0 < tags.length )
            tags = [@"#" stringByAppendingString:tags];
        else if( 0 == txt.length )
            tags = self.shaarli.tagsDefault;
        else
            tags = @"";
        if( 0 < tags.length )
            tags = [tags stringByAppendingString:@" "];

        txt = [tags stringByAppendingString:txt];
    }
    self.textView.text = txt;
}


-(void)didPostLoginForm:(NSMutableDictionary *)form toURL:(NSURL *)dst error:(NSError *)error
{
    MRLogD(@"%@ %@", form, error, nil);
    if( error ) {
        [self cancelWithError:error];
        return;
    }
    self.postForm = form;
    self.postURL = dst;
    if( !self.wasTouched )
        [self form2Gui:form];
    else {
        MRLogD(@"Evtl. show a hint that there was discarded incoming server data.", nil);
    }
}


-(void)didFinishPostFormToURL:(NSURL *)dst error:(NSError *)error
{
    MRLogD(@"%@ %@", dst, error, nil);
    if( error ) {
        [self cancelWithError:error];
        return;
    }
    // http://www.pixeldock.com/blog/ios8-share-extension-completionhandler-for-loaditemfortypeidentifier-is-never-called/
    // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
    [super didSelectPost];
    self.post = nil;
}


@end

File: swift4/Share/ShareVC.swift
//
//  ShareVC.swift
//  Share
//
//  Created by Marcus Rohrmoser on 02.03.20.
//  Copyright © 2020-2021 Marcus Rohrmoser mobile Software http://mro.name/me. All rights reserved.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import UIKit
import Social
import MobileCoreServices
import AudioToolbox

private func stringFromPrivacy(_ priv : Bool) -> String {
    return priv
        ? NSLocalizedString("Private 🔐", comment:"ShareVC")
        : NSLocalizedString("Public 🔓", comment:"ShareVC")
}

private func privacyFromString(_ s : String) -> Bool {
    return s != stringFromPrivacy(false)
}

private func play_sound_ok() {
    // https://github.com/irccloud/ios/blob/6e3255eab82be047be141ced6e482ead5ac413f4/ShareExtension/ShareViewController.m#L155
    AudioServicesPlaySystemSound(1001)
}

private func play_sound_err() {
    AudioServicesPlayAlertSound(kSystemSoundID_Vibrate)
}

@objc (ShareVC) // https://blog.hellocode.co/post/share-extension/
class ShareVC: SLComposeServiceViewController {

    private let semver          =  info_to_semver(Bundle.main.infoDictionary)

    private var current         : BlogM? // may become local if not for missing cfg error
    private var wasTouched      = false
    private var itemTitle       : SLComposeSheetConfigurationItem?
    private var itemAudience    : SLComposeSheetConfigurationItem?
    private var session         : URLSession?
    private var action          : URL = URLEmpty
    private var ctx             : HtmlFormDict = [:]
    private var url             : URL = URLEmpty

    override func viewDidLoad() {
        debugPrint("viewDidLoad")
        super.viewDidLoad()
    }

    override func configurationItems() -> [Any]! {
        debugPrint("configurationItems")
        guard let iTi = SLComposeSheetConfigurationItem() else {return []}
        iTi.title = NSLocalizedString("Title", comment:"ShareVC")
        iTi.value = self.contentText

        guard let iAu = SLComposeSheetConfigurationItem() else {return []}
        iAu.title = NSLocalizedString("Audience", comment:"ShareVC")
        iAu.value = stringFromPrivacy(false)
        weak var wself = self
        iAu.tapHandler = {
            guard let sf = wself else {return}
            guard let iAu = sf.itemAudience else {return}
            iAu.value = stringFromPrivacy( !privacyFromString(iAu.value) )
            sf.wasTouched = true
        }

        itemTitle = iTi
        itemAudience = iAu
        return [iTi, iAu]
    }

    override func viewWillAppear(_ animated: Bool) {
        debugPrint("viewWillAppear")
        super.viewWillAppear(animated)
        view.tintColor = UIColor(red:128 / 255.0, green:173 / 255.0, blue:72 / 255.0, alpha:1.0)
        assert(itemTitle != nil)
        assert(itemAudience != nil)

        let sha = ShaarliM.shared
        // sha.defaults.removePersistentDomain(forName:"group.\(BUNDLE_ID)") // doesn't do it.
        current = sha.loadBlog(sha.defaults)
        guard let current = current else {
            // do nothing here and let viewDidAppear display a error popup
            return
        }
        let cli = ShaarliHtmlClient(semver)

        textView.keyboardType = .twitter
        view.subviews.forEach({ (v) in
            // dark mode?
            v.backgroundColor = UIColor.white.withAlphaComponent(0.89)
        })

        guard let itemTitle = itemTitle else {return}
        guard let itemAudience = itemAudience else {return}
        guard let textView = textView else {return}

        title = current.title
        itemTitle.value = contentText
        let preset = tagsNormalise(description:itemTitle.value, extended:current.tagsDefault, tags:[], known:[])
        textView.text = "\(preset.extended) \(NSLocalizedString("🔄", comment:"ShareVC"))"
        itemAudience.value = stringFromPrivacy(current.privateDefault)

        let tPli = kUTTypePropertyList as String
        let tUrl = kUTTypeURL as String
        let tTxt = kUTTypeText as String
        let RK = NSExtensionJavaScriptPreprocessingResultsKey as String
        weak var ws = self
        for _item in (extensionContext?.inputItems)! {
            let item = _item as! NSExtensionItem
            for ip in (item.attachments!) {
                guard let ws = ws else {return}
                // let ip = _ip as! NSItemProvider // required for Xcode <10
                if( ip.hasItemConformingToTypeIdentifier(tPli)) {
                    ip.loadItem(forTypeIdentifier:tPli, options:nil) { (_dic, err) in
                        guard let _dic = (_dic as? NSDictionary)?[RK] as? [String:String] else {
                            play_sound_err()
                            ws.showError(
                                title:NSLocalizedString("URL Share Sheet failed", comment: "ShareVC"),
                                message:NSLocalizedString("I got no dictionary to share.", comment: "ShareVC"),
                                showsettings:false
                            )
                            return
                        }
                        let kUrl = "url"
                        let kTit = "title"
                        let kDsc = "description"
                        let kTgs = "keywords"
                        let kImg = "image"
                        guard let _url = URL(string:_dic[kUrl] ?? "") else {
                            play_sound_err()
                            ws.showError(
                                title:NSLocalizedString("URL Share Sheet failed", comment: "ShareVC"),
                                message:NSLocalizedString("I got no url to share.", comment: "ShareVC"),
                                showsettings:false
                            )
                            return
                        }
                        guard let err = err else {
                            let t0 = Date()
                            cli.get(current.endpoint, current.credential, current.timeout, _url, { (ses, act, ctx, _url, tit, dsc, tgs, pri, tim, seti, err) in
                                guard "" == err else {
                                    play_sound_err()
                                    ws.showError(
                                        title:NSLocalizedString("Can't reach Shaarli", comment: "ShareVC"),
                                        message:err,
                                        showsettings:true
                                    )
                                    return
                                }
                                guard URLEmpty != act else {
                                    play_sound_err()
                                    ws.showError(
                                        title:NSLocalizedString("Can't post to Shaarli", comment: "ShareVC"),
                                        message:NSLocalizedString("the Shaarli responded an empty linkform action, I don't know where to post to.", comment: "ShareVC"),
                                        showsettings:true
                                    )
                                    return
                                }
                                self.session = ses
                                // should 'old' come out of get( callback(... old) )?
                                let old = isOld(t0, seti, cli.timeShaarli(current.timezone, tim))
                                let dti = _dic[kTit] ?? ""
                                let dde = _dic[kDsc] ?? ""
                                // let dim = URL(string:_dic[kImg] ?? "", relativeTo:_url)
                                let v = old || ("" != tit || "" != dsc)
                                    ? (tit, dsc, tgs)
                                    : (dti == "" && dde == ""
                                        ? (itemTitle.value ?? "", preset.extended, [])
                                        : (dti, dde, tagsSplit(_dic[kTgs])) )
                                let r = tagsNormalise(description:v.0, extended:v.1, tags:v.2.union(preset.tags), known:[])
                                DispatchQueue.main.async {
                                    ws.action = act
                                    ws.ctx = ctx
                                    ws.url = _url
                                    itemTitle.value = r.description
                                    textView.text = "\(r.extended) "
                                    itemAudience.value = stringFromPrivacy(pri)
                                }
                            })
                            return
                        }
                        ws.showError(
                            title:NSLocalizedString("URL Share Sheet failed", comment: "ShareVC"),
                            message:err.localizedDescription,
                            showsettings:false
                        )
                    }
                }
                // see predicate from http://stackoverflow.com/a/27932776
                else if( ip.hasItemConformingToTypeIdentifier(tUrl) ) {
                    // maybe this whole block can go away
                    ip.loadItem(forTypeIdentifier:tUrl, options:nil) { (_url, err) in
                        guard let _url = _url as? URL else {
                            play_sound_err()
                            ws.showError(
                                title:NSLocalizedString("URL Share Sheet failed", comment: "ShareVC"),
                                message:NSLocalizedString("I got no url to share.", comment: "ShareVC"),
                                showsettings:false
                            )
                            return
                        }
                        guard let err = err else {
                            cli.get(current.endpoint, current.credential, current.timeout, _url, { (ses, act, ctx, _url, tit, dsc, tgs, pri, tim, seti, err) in
                                guard "" == err else {
                                    play_sound_err()
                                    ws.showError(
                                        title:NSLocalizedString("Can't reach Shaarli", comment: "ShareVC"),
                                        message:err,
                                        showsettings:true
                                    )
                                    return
                                }
                                guard URLEmpty != act else {
                                    play_sound_err()
                                    ws.showError(
                                        title:NSLocalizedString("Can't post to Shaarli", comment: "ShareVC"),
                                        message:NSLocalizedString("the Shaarli responded an empty linkform action, I don't know where to post to.", comment: "ShareVC"),
                                        showsettings:true
                                    )
                                    return
                                }
                                self.session = ses
                                let v = "" == tit && "" == dsc
                                    ? (itemTitle.value, preset.extended)
                                    : (tit, dsc)
                                let r = tagsNormalise(description:v.0, extended:v.1, tags:tgs.union(preset.tags), known:[])
                                DispatchQueue.main.async {
                                    ws.action = act
                                    ws.ctx = ctx
                                    ws.url = _url
                                    itemTitle.value = r.description
                                    textView.text = "\(r.extended) "
                                    itemAudience.value = stringFromPrivacy(pri)
                                }
                            })
                            return
                        }
                        ws.showError(
                            title:NSLocalizedString("URL Share Sheet failed", comment: "ShareVC"),
                            message:err.localizedDescription,
                            showsettings:false
                        )
                    }
                }
                else if( ip.hasItemConformingToTypeIdentifier(tTxt) ) {
                    ip.loadItem(forTypeIdentifier:tTxt, options:nil) { (_txt, err) in
                        guard let err = err else {
                            debugPrint("done. title:\(itemTitle.value ?? "-") txt:\(String(describing: _txt))")
                            return
                        }
                        ws.showError(
                            title:NSLocalizedString("TXT Share Sheet failed", comment: "ShareVC"),
                            message:err.localizedDescription,
                            showsettings:false
                        )
                    }
                }
            }
        }
    }

    override func didSelectPost() {
        debugPrint("didSelectPost")
        let c = ShaarliHtmlClient(semver)
        guard let tit = itemTitle?.value else {return}
        guard let dsc = textView.text else {return}
        let pri = privacyFromString((itemAudience?.value)!)
        let r = tagsNormalise(description:tit, extended:dsc, tags:[], known:[])
        c.add(session!, action, ctx, url, r.description, r.extended, r.tags, pri) { err in
            guard "" == err else {
                play_sound_err()
                self.showError(
                    title:NSLocalizedString("Share failed", comment: "ShareVC"),
                    message:err,
                    showsettings:false
                )
                usleep(750 * 1000)
                return
            }

            play_sound_ok()
            // wait until the sound finished
            usleep(750 * 1000)
            super.didSelectPost()
        }
        // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
        // self.extensionContext!.completeRequest(returningItems: [], completionHandler: nil)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard nil != current else {
            showError(
                title:NSLocalizedString("No Shaarli found", comment:"ShareVC"),
                message:NSLocalizedString("Please add one in the Settings.", comment:"ShareVC"),
                showsettings:true)
            return
        }

        wasTouched = false
    }

    private func showError(title:String, message:String, showsettings:Bool) {
        DispatchQueue.main.async {
            let alert = UIAlertController(
                title:title,
                message:message,
                preferredStyle:.alert
            )

            alert.addAction(UIAlertAction(
                title: NSLocalizedString("Cancel", comment:"ShareVC"),
                style:.cancel,
                handler:{ (_) in
                    self.cancel()
            }))

            if showsettings {
                alert.addAction(UIAlertAction(
                    title: NSLocalizedString("Settings", comment:"ShareVC"),
                    style:.default,
                    handler:{ (_) in
                        // https://stackoverflow.com/a/44499222/349514
                        DispatchGroup().notify(queue: DispatchQueue.main) {
                            let _ = self.openURL(URL(string:"\(SELF_URL_PREFIX):///settings")!)
                        }
                        self.cancel()
                }))
            }

            self.present(alert, animated:true, completion:nil)
        }
    }

    override func presentationAnimationDidFinish() {
        debugPrint("presentationAnimationDidFinish")
    }

    override func isContentValid() -> Bool {
        debugPrint("isContentValid")
        // Do validation of contentText and/or NSExtensionContext attachments here
        wasTouched = true
        return true
    }

    // No preview image right upper inside the share dialog.
    override func loadPreviewView() -> UIView! {
        return nil
    }

    override func didSelectCancel() {
        debugPrint("didSelectCancel")
        super.didSelectCancel()
    }

    // https://stackoverflow.com/a/44499222/349514
    // Function must be named exactly like this so a selector can be found by the compiler!
    // Anyway - it's another selector in another instance that would be "performed" instead.
    @objc private func openURL(_ url: URL) -> Bool {
        var rep: UIResponder? = self
        while rep != nil {
            if let app = rep as? UIApplication {
                return app.perform(#selector(openURL(_:)), with: url) != nil
            }
            rep = rep?.next
        }
        return false
    }
}
